keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -51,223 +48,183 @@
+  Based on [the heuristics described above](#SELinux-support-in-volumes), it may pass `:Z` option to pod volumes, the container runtime allocates a new SELinux context for the pod and relabel all the volumes recursively.
+
+* When kubelet knows that the volume does not support mounting with `-o context` (or when it is not sure), it falls back to the current behavior.
+  Based on [the heuristics described above](#SELinux-support-in-volumes), it may pass `:Z` option to pod volumes, and the container will relabel the volume recursively.
+
+**How kubelet knows *if* a volume supports mounting with SELinux?**
+
+[As described above](#volume-mounting), any volume can be mounted with `-o context=XYZ`, as long as it's the first mount of the volume.
+
+* For mounting of block devices, it's safe to assume that `-o context=XYZ` always works - kubelet / CSI drivers mounts such a volume as whole.
+* For shared volumes, such as NFS, GlusterFS or CephFS, it's hard for kubelet to distinguish what is a volume and what is a directory on it.
+  For example, when mounting `example.com:/exports/archive/jsafrane/projects/foo`, `example.com:/exports/archive` is name of the volume (from kernel point of view) and `jsafrane/projects/foo` is path inside the volume.
+  It can still be mounted with `-o context=XYZ`, however, all subsequent mounts of `example.com:/exports/archive` must use the same `-o context=XYZ` option!
+  In Kubernetes, we do not have any way to tell if a NFS, GlusterFS or CephFS PVs are using the same volumes (NFS/Gluster/CephFS shares), and therefore kubelet cannot use `-o context=XYZ` for these volume plugins.
+
+* For CSI drivers, kubelet has no clue if the PVs provided by the driver are independent and each of them can be mounted with a different `-o context=` value, or some (or all) PVs are in fact the same volume from kernel point of view, only a different subdirectory of it.
+    Therefore, we need CSI driver vendors to provide such information in CSIDriver object of their driver:
+
+    ```go
+    // In storage.k8s.io/v1:
+    
+    
+    // CSIDriverSpec is the specification of a CSIDriver.
+    type CSIDriverSpec struct {
+        // SELinuxMountSupported specifies if the CSI driver supports "-o context"
+        // mount option.
+        //
+        // When "true", Kubernetes may call NodeStage / NodePublish with "-o context=xyz" mount
+        // option. The CSI driver must ensure that all volumes can be mounted with different
+        // `-o context` options. This is typical for storage backends that provide volumes
+        // as filesystems on block devices or as independent shared volumes.
+        //
+        // When "false", Kubernetes won't pass any special SELinux mount options to the driver.
+        // This is typical for volumes that represent subdirectories of a bigger shared filesystem.
+        //
+        // Default is "false".
+        SELinuxMountSupported *bool;
```

Comment:
This is a setting at CSI driver level.  Do we need something that has a finer granularity?  For example, what if a CSI driver supports both block and file share volumes and it is true for block but false for file share volumes?

Reply:
We would need CSI change for this - the driver would need to advertise which volume supports SELinux mount options and which does not. Kubelet cannot not know that. I tried to avoid CSI changes here.

Reply:
A single CSI driver supporting multiple volume types is not well supported today. I think a CSI driver will have to filter out the SELinux mount option in the NodeStage/Publish call.

Reply:
*If* we expect CSI drivers to filter SELinux options, then I'd prefer to have an explicit field in CSI NodeStage/NodePublish for that, so they know what to filter.

Do we really expect that? I am afraid that the most CSI drivers won't care about SELinux at all, as it's used in minority of Kubernetes distros.

Reply:
I think in the onprem world there are more users that care about running on RHEL with SELinux. Maybe leave this as a open item to resolve before beta?